src/pybind/mgr/nfs/export.py

```diff
@@ -497,46 +519,44 @@ def get_export_by_pseudo(
         export = self._fetch_export(cluster_id, pseudo_path)
         return export.to_dict() if export else None
 
-    def apply_export(self, cluster_id: str, export_config: str) -> Tuple[int, str, str]:
+    def apply_export(self, cluster_id: str, export_config: str) -> AppliedExportResults:
         try:
-            if not export_config:
-                raise NFSInvalidOperation("Empty Config!!")
-            try:
-                j = json.loads(export_config)
-            except ValueError:
-                # okay, not JSON.  is it an EXPORT block?
+            aeresults = AppliedExportResults()
+            for export in self._read_export_config(cluster_id, export_config):
                 try:
-                    blocks = GaneshaConfParser(export_config).parse()
-                    exports = [
-                        Export.from_export_block(block, cluster_id)
-                        for block in blocks
-                    ]
-                    j = [export.to_dict() for export in exports]
+                    chg = self._apply_export(cluster_id, export)
                 except Exception as ex:
-                    raise NFSInvalidOperation(f"Input must be JSON or a ganesha EXPORT block: {ex}")
-
-            # check export type
-            if isinstance(j, list):
-                ret, out, err = (0, '', '')
-                for export in j:
-                    try:
-                        r, o, e = self._apply_export(cluster_id, export)
-                    except Exception as ex:
-                        r, o, e = exception_handler(ex, f'Failed to apply export: {ex}')
-                        if r:
-                            ret = r
-                    if o:
-                        out += o + '\n'
-                    if e:
-                        err += e + '\n'
-                return ret, out, err
-            else:
-                r, o, e = self._apply_export(cluster_id, j)
-                return r, o, e
+                    msg = f'Failed to apply export: {ex}'
+                    log.exception(msg)
+                    chg = {"state": "error", "msg": msg}
+                aeresults.append(chg)
+            return aeresults
         except NotImplementedError:
```

Comment:
Just repeating what I said in https://github.com/ceph/ceph/pull/46209/commits/6b6c2d5d85e6101e3ea755bd8b1d10e45a5cc525#r946138712 in case it gets missed. 

The NotImplementedError exception raised by _apply_export() will be caught by the inner catch all exception. I'm not sure if it will be propagated here.

Reply:
FWIW I think this confusion existed in the earlier code. If I'm correct, the original function would never raise NotImplementedError in the first half, the part I turned into _read_export_config, but only in the part that executed _apply_export which did also have an "inner try-except block". I'm happy to fix it whichever way you prefer (choices in my previous comment).

Reply:
Yeah, that's my understanding too, only _apply_export()  can raise NotImplementedError.

How about handling the NotImplementedError within _apply_export() and that would simplify apply_export()
```  
def _apply_export(self, cluster_id: str, new_export_dict: Dict) -> Dict[str, str]:
       ...
       try:
           self._update_export(cluster_id, new_export, need_nfs_service_restart)
       # your comment about NotImplementedError    
       except NotImplementedError:
            return {"pseudo": new_export.pseudo, "state": "changes applied (Manual restart of NFS Pods required)"}
             # My concern is that the user could miss reading the important bit about the restart.
        return {"pseudo": new_export.pseudo, "state": "updated"}

def apply_export(self, cluster_id: str, export_config: str) -> AppliedExportResults:   
        try:
            exports = self._read_export_config(cluster_id, export_config)
        except Exception as e:
            log.exception(f'Failed to update export: {e}')
            raise ErrorResponse.wrap(e)
        
        aeresults = AppliedExportResults()
        for export in exports:
            try:
                chg = self._apply_export(cluster_id, export)
            except Exception as ex:
                msg = f'Failed to apply export: {ex}'
                log.exception(msg)
                chg = {"state": "error", "msg": msg}
                aeresults.append(chg)
        return aeresults 
```

I wish the interface was designed to allow only applying a single export.

Reply:
I think that sounds reasonable. Will do.

Reply:
> I wish the interface was designed to allow only applying a single export.

Big agreement here.

Reply:
I implemented a hybrid of your suggestion and one of mine. While I think my analysis is correct that NotImplmementedError is usually raised by the _update_export function, I'm not entirely sure that's the only place, so I ended up adding an intermediate function that handles the NotImplementedError from _apply_export as well as the general error case.  I hope you think it's OK. If you strongly prefer I only handle the NotImplmementedError from  _update_export - let me know I'll do so. But this makes me a bit happier that all possible cases are "protected".

Reply:
I like the changes. Thanks!